Support cloud storage sink (#10998)
pingcap/docs: ticdc/ticdc-sink-to-cloud-storage.md

- `--sink-uri`: The downstream address of the changefeed. For details, see [Configure Cloud Storage in Sink URI](#configure-sink-uri).
- `--start-ts`: The starting TSO of the changefeed. From this TSO, the TiCDC cluster starts pulling data. The default value is the current time.
- `--target-ts`: The ending TSO of the changefeed. To this TSO, the TiCDC cluster stops pulling data. The default value is empty, which means that TiCDC does not automatically stop pulling data.
- `--config`: The configuration file of the `changefeed`. For details, see [TiCDC changefeed configuration parameters](/ticdc/ticdc-changefeed-config.md)

## Configure sink URI

This section describes how to configure cloud storage, including `S3`, `Azure Blob Storage`, and `NFS`, in the changefeed URI.

### Configure S3 or Azure Blob Storage in sink URI

The URI parameters of Amazon S3 and Azure Blob Storage in TiCDC are the same as their URL parameters in BR. For details, see [Backup storage URL format](/br/backup-and-restore-storages#url-format-description).

### Configure NFS in sink URI

The following configuration saves row change events to NFS:

```
shell
--sink-uri="file:///my-directory/prefix"
```

> Other parameters optional in the URI are as follows:

Reviewer: ```suggestion
Other optional parameters in the URI are as follows:
```

| Parameter         | Description                                             |
| :------------ | :------------------------------------------------ |
| `worker-count` | Concurrency for saving data changes to cloud storage in the downstream (optional, default value: `16`, value range: [`1`, `512`] |
| `flush-interval` | Interval for saving data changes to cloud storage in the downstream (optional, default value: `5s`, value range: [`2s`, `10m`] |

> | `file-size` | A data change file is stored to cloud storage if its bytes exceed the value of this parameter (optional, default value: `67108864`, value range: [`67108864`, `536870912`]) |

Reviewer: 中文的取值范围下限是 1048576
```suggestion
| `file-size` | A data change file is stored to cloud storage if its bytes exceed the value of this parameter (optional, default value: `67108864`, value range: [`1048576`, `536870912`]) |
```

## Storage path structure

This section describes the storage path structure of data change records, metadata, and DDL events.

### Data change records

Data change records are saved to the following path:

```shell
{protocol}://{prefix}/{schema}/{table}/{table-version-separator}/{partition-separator}/{date-separator}/CDC{num}.{extension}
```

- `protocol`: specifies the data transmission protocol, or the storage type, for example, `s3://xxxxx`.
- `prefix`: specifies the user-defined parent directory, for example, `s3://bucket/bbb/ccc`.
- `schema`: specifies the schema name, for example, `s3://bucket/bbb/ccc/test`.
- `table`: specifies the table name, for example, `s3://bucket/bbb/ccc/test/table1`.
- `table-version-separator`: specifies the separator that separates the file path by the table version, for example, `s3://bucket/bbb/ccc/test/table1/9999`.
- `partition-separator`: specifies the separator that separates the file path by the table partition, for example, `s3://bucket/bbb/ccc/test/table1/9999/20`
- `date-separator`: classifies the files by the transaction commit date. Value options are:

>     - none: no date-separator. For example, all files with `test.table1` version being `9999` are saved to `s3://bucket/bbb/ccc/test/table1/9999`.

Reviewer: ```suggestion
    - `none`: no date-separator. For example, all files with `test.table1` version being `9999` are saved to `s3://bucket/bbb/ccc/test/table1/9999`.
```

    - `year`: the separator is the year a transaction is committed, for example, `s3://bucket/bbb/ccc/test/table1/9999/2022`.
    - `month`: the separator is the year and month a transaction is committed, for example,`s3://bucket/bbb/ccc/test/table1/9999/2022-01`.
    - `day`: the separator is the year, month, and day a transaction is committed, for example,`s3://bucket/bbb/ccc/test/table1/9999/2022-01-02`.
- `num`: saves the serial number of the file that records the data change, for example, `s3://bucket/bbb/ccc/test/table1/9999/2022-01-02/CDC000005.csv`.

> - `extension`: specifies the extension of the file. TiDB v6.5.0 supports the CSV format only.

Reviewer: 中文版说支持两种格式。
```suggestion
- `extension`: specifies the extension of the file. TiDB v6.5.0 supports the CSV and canal-json formats only.
```

            "ColumnType":"VARCHAR",
            "ColumnLength":"30"
        },
        {
            "ColumnName":"HireDate",
            "ColumnType":"DATETIME"
        },
        {
            "ColumnName":"OfficeLocation",
            "ColumnType":"BLOB",
            "ColumnLength":"20"
        }
    ],
    "TableColumnsTotal":"5"
}
```

- `Table`: Table name.
- `Schema`: Schema name.
- `Version`: Table version.
- `TableColumns`: An array of one or more maps, each of which describes a column in the source table.
    - `ColumnName`: Column name.
    - `ColumnType`: Column type. For details, see [Data type](#data-type).
    - `ColumnLength`: Column length. For details, see [Data type](#data-type).
    - `ColumnPrecision`: Column precision. For details, see [Data type](#data-type).
    - `ColumnScale`: The number of digits following the decimal point (the scale). For details, see [Data type](#data-type).
    - `ColumnNullable`: The column can be null when the value of this option is `true`.
    - `ColumnIsPk`: The column is part of the primary key when the value of this option is `true`.
- `TableColumnsTotal`: The size of the `TableColumns` array.

### Data type

> Data type is defined as `T(M[, D])`. For details, see [Data Types](/data-type-overview.md).

Reviewer: ```suggestion
The data type is defined as `T(M[, D])`. For details, see [Data Types](/data-type-overview.md).
```

Decimal types are defined as follows in `schema.json`:

```shell
{
    "ColumnName":"COL1",
    "ColumnType":"{DT} [UNSIGNED]",
    "ColumnPrecision":"{M}",
    "ColumnScale":"{D}"
}
```

#### Date and time types

Date types in TiDB are defined as `DT`, where

- `DT` is the date type, which can be `DATE` or `YEAR`.

The date types are defined as follows in `schema.json`:

```shell
{
    "ColumnName":"COL1",
    "ColumnType":"{DT}"
}
```

The time types in TiDB are defined as `TT[(M)]`, where

- `TT` is the time type, which can be `TIME`, `DATETIME`, or `TIMESTAMP`.
- `M` is the precision of seconds in the range from 0 to 6.

The time types are defined as follows in `schema.json`:

```shell
{
    "ColumnName":"COL1",
    "ColumnType":"{TT}",
    "ColumnScale":"{M}"
}
```

#### String types

The string types in TiDB are defined as `ST[(M)]`, where

- `ST` is the string type, which can be `CHAR`, `VARCHAR`, `TEXT`, `BINARY`, `BLOB`, or `JSON`.
- `M` is the maximum length of the string.

The string types are defined as follows in `schema.json`:

```shell
{
    "ColumnName":"COL1",
    "ColumnType":"{ST}",
    "ColumnLength":"{M}"
}
```

> #### Enum and Set types

Reviewer: 代码示例中用的是 ENUM/SET，保持一致吧。
```suggestion
#### ENUM/SET types
```

Author: 但其实是两种类型，见 [`ENUM` type](https://docs.pingcap.com/tidb/dev/data-type-string#enum-type) 和 [`SET` type](https://docs.pingcap.com/tidb/dev/data-type-string#set-type)，我改成大写然后 and 连接吧，鉴于文档中不推荐用 / 表示并列关系。

- `date-separator`: classifies the files by the transaction commit date. Value options are:
    - `none`: no `date-separator`. For example, all files with `test.table1` version being `9999` are saved to `s3://bucket/bbb/ccc/test/table1/9999`.
    - `year`: the separator is the year a transaction is committed, for example, <code>s3://bucket/bbb/ccc/test/table1/9999/**2022**</code>.
    - `month`: the separator is the year and month a transaction is committed, for example, <code>s3://bucket/bbb/ccc/test/table1/9999/**2022-01**</code>.
    - `day`: the separator is the year, month, and day a transaction is committed, for example, <code>s3://bucket/bbb/ccc/test/table1/9999/**2022-01-02**</code>.
- `num`: saves the serial number of the file that records the data change, for example, <code>s3://bucket/bbb/ccc/test/table1/9999/2022-01-02/CDC**000005**.csv</code>.
- `extension`: specifies the extension of the file. TiDB v6.5.0 supports the CSV and canal-json formats.

> **Note:**
>
> The table version changes in the following two cases:
>

> > - After DDL operations are performed, the table version is the TSO when the DDL is executed in the upstream TiDB. However, the change of the table version does not mean the change of the table schema. For example, adding a comment to a column does not cause the `schema.json` file content to change.

Reviewer 2: ```suggestion
> - After a DDL operation is performed, the table version is the TSO when the DDL is executed in the upstream TiDB. However, the change of the table version does not mean the change of the table schema. For example, adding a comment to a column does not cause the `schema.json` file content to change.
```

> - The changefeed process restarts. The table version is the checkpoint TSO when the process restarts. When there are many tables and the process restarts, it takes a long time to traverse all directories and find the position where each table was written last time. Therefore, data is written to a new directory with the version being the checkpoint TSO, instead of to the earlier directory.

### Metadata

Metadata is saved in the following path:

```shell
{protocol}://{prefix}/metadata
```

Metadata is a JSON-formatted file, for example:

> ```shell

Reviewer 2: ```suggestion
```json
```

{
    "checkpoint-ts":433305438660591626
}
```

- `checkpoint-ts`: Transactions with `commit-ts` smaller than `checkpoint-ts` are written to the target storage in the downstream.

### DDL events

> When DDL events cause the table version to change, TiCDC switches to the new path to write data change records. For example, when the version of `test.table1` changes from `9999` to `10000`, data will be written to the path `s3://bucket/bbb/ccc/test/table1/10000/2022-01-02/CDC000001.csv`. In addition, when DDL events occur, TiCDC generates a `schema.json` file to save the table schema information.

Reviewer 2: ```suggestion
When DDL events cause the table version to change, TiCDC switches to a new path to write data change records. For example, when the version of `test.table1` changes from `9999` to `10000`, data will be written to the path `s3://bucket/bbb/ccc/test/table1/10000/2022-01-02/CDC000001.csv`. In addition, when DDL events occur, TiCDC generates a `schema.json` file to save the table schema information.
```

> Table  schema information is saved in the following path:

Reviewer 2: ```suggestion
Table schema information is saved in the following path:
```

        },
        {
            "ColumnName":"LastName",
            "ColumnType":"CHAR",
            "ColumnLength":"20"
        },
        {
            "ColumnName":"FirstName",
            "ColumnType":"VARCHAR",
            "ColumnLength":"30"
        },
        {
            "ColumnName":"HireDate",
            "ColumnType":"DATETIME"
        },
        {
            "ColumnName":"OfficeLocation",
            "ColumnType":"BLOB",
            "ColumnLength":"20"
        }
    ],
    "TableColumnsTotal":"5"
}
```

- `Table`: Table name.
- `Schema`: Schema name.
- `Version`: Protocol version of the storage sink.
- `TableVersion`: Table version.
- `Query`：DDL statement.
- `TableColumns`: An array of one or more maps, each of which describes a column in the source table.
    - `ColumnName`: Column name.
    - `ColumnType`: Column type. For details, see [Data type](#data-type).
    - `ColumnLength`: Column length. For details, see [Data type](#data-type).
    - `ColumnPrecision`: Column precision. For details, see [Data type](#data-type).
    - `ColumnScale`: The number of digits following the decimal point (the scale). For details, see [Data type](#data-type).

>     - `ColumnNullable`: The column can be null when the value of this option is `true`.

Reviewer 2: ```suggestion
    - `ColumnNullable`: The column can be NULL when the value of this option is `true`.
```

>     - `ColumnNullable`: The column can be null when the value of this option is `true`.

Reviewer 2: ```suggestion
    - `ColumnNullable`: The column can be NULL when the value of this option is `true`.
```

## Configure sink URI

This section describes how to configure cloud storage, including `S3`, `Azure Blob Storage`, and `NFS`, in the changefeed URI.

### Configure S3 or Azure Blob Storage in sink URI

The URI parameters of Amazon S3 and Azure Blob Storage in TiCDC are the same as their URL parameters in BR. For details, see [Backup storage URL format](/br/backup-and-restore-storages.md#url-format-description).

### Configure NFS in sink URI

The following configuration saves row change events to NFS:

```
shell
--sink-uri="file:///my-directory/prefix"
```

Other optional parameters in the URI are as follows:

| Parameter         | Description                                             |
| :------------ | :------------------------------------------------ |
| `worker-count` | Concurrency for saving data changes to cloud storage in the downstream (optional, default value: `16`, value range: [`1`, `512`] |
| `flush-interval` | Interval for saving data changes to cloud storage in the downstream (optional, default value: `5s`, value range: [`2s`, `10m`] |

> | `file-size` | A data change file is stored to cloud storage if its bytes exceed the value of this parameter (optional, default value: `67108864`, value range: [`1048576`, `536870912`]) |

Reviewer 3: ```suggestion
| `file-size` | A data change file is stored to cloud storage if the number of bytes exceed the value of this parameter (optional, default value: `67108864`, value range: [`1048576`, `536870912`]) |
```

---
title: Replicate Data to Storage Services
summary: Learn how to replicate data to storage services using TiCDC, and learn about the storage path of the replicated data.
---

# Replicate Data to Storage Services

> **Warning:**
>

> > This feature is experimental. It is not recommended to use it in the production environment.

Author: IMO, the original is more frequently used for experimental features, and `the current version` may make it hard to maintain.

```

- `--changefeed-id`: The ID of the changefeed. The format must match the `^[a-zA-Z0-9]+(\-[a-zA-Z0-9]+)*$` regular expression. If this ID is not specified, TiCDC automatically generates a UUID (the version 4 format) as the ID.
- `--sink-uri`: The downstream address of the changefeed. For details, see [Configure sink URI](#configure-sink-uri).
- `--start-ts`: The starting TSO of the changefeed. TiCDC starts pulling data from this TSO. The default value is the current time.
- `--target-ts`: The ending TSO of the changefeed. TiCDC stops pulling data until this TSO. The default value is empty, which means that TiCDC does not automatically stop pulling data.
- `--config`: The configuration file of the changefeed. For details, see [TiCDC changefeed configuration parameters](/ticdc/ticdc-changefeed-config.md).

## Configure sink URI

This section describes how to configure storage services in the changefeed URI, including Amazon S3, Azure Blob Storage, and NFS.

> ### Configure S3 or Azure Blob Storage in sink URI

Reviewer 2: ```suggestion
### Configure Amazon S3 or Azure Blob Storage
```

The URI parameters of Amazon S3 and Azure Blob Storage in TiCDC are the same as their URL parameters in BR. For details, see [Backup storage URL format](/br/backup-and-restore-storages.md#url-format-description).

> ### Configure NFS in sink URI

Reviewer 2: ```suggestion
### Configure NFS
```

Decimal types are defined as follows in `schema.json`:

```json
{
    "ColumnName":"COL1",
    "ColumnType":"{DT} [UNSIGNED]",
    "ColumnPrecision":"{M}",
    "ColumnScale":"{D}"
}
```

#### Date and time types

Date types in TiDB are defined as `DT`, where

- `DT` is the date type, which can be `DATE` or `YEAR`.

The date types are defined as follows in `schema.json`:

```json
{
    "ColumnName":"COL1",
    "ColumnType":"{DT}"
}
```

The time types in TiDB are defined as `TT[(M)]`, where

- `TT` is the time type, which can be `TIME`, `DATETIME`, or `TIMESTAMP`.
- `M` is the precision of seconds in the range from 0 to 6.

The time types are defined as follows in `schema.json`:

```json
{
    "ColumnName":"COL1",
    "ColumnType":"{TT}",
    "ColumnScale":"{M}"
}
```

#### String types

The string types in TiDB are defined as `ST[(M)]`, where

- `ST` is the string type, which can be `CHAR`, `VARCHAR`, `TEXT`, `BINARY`, `BLOB`, or `JSON`.
- `M` is the maximum length of the string.

The string types are defined as follows in `schema.json`:

```json
{
    "ColumnName":"COL1",
    "ColumnType":"{ST}",
    "ColumnLength":"{M}"
}
```

#### Enum and Set types

> The `ENUM` and `SET` types are defined as follows in `schema.json`:

Reviewer 2: ```suggestion
The Enum and Set types are defined as follows in `schema.json`:
```
